File: java/src/main/java/com/example/vulnapp/vulns/SqlInjectionController.java
package com.example.vulnapp.vulns;

import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * SQL Injection — OWASP 2021 A03 Injection | OWASP 2025 A05 Injection.
 *
 * Reference module: the comment/annotation format used here is mirrored by
 * every other vulnerability controller in this app. Each permutation carries a
 * standard header describing the flaw, its CWE, an exploitation-likelihood
 * rating, an example payload, and the corresponding safe pattern.
 */
@RestController
@RequestMapping("/injection/sql")
public class SqlInjectionController {

    private final JdbcTemplate jdbc;

    public SqlInjectionController(JdbcTemplate jdbc) {
        this.jdbc = jdbc;
    }

    // ========================================================================
    // PERMUTATION 1 — SQL injection via string concatenation (auth bypass)
    // OWASP 2021: A03 Injection      OWASP 2025: A05 Injection
    // CWE-89: Improper Neutralization of Special Elements used in an SQL Command
    // EXPLOITATION LIKELIHOOD: HIGH
    //   Reachable pre-authentication, deterministic, no tooling required.
    //   The textbook  ' OR '1'='1  / username=admin'--  payloads bypass login.
    // PREVALENCE TODAY: greenfield LOW-MED (modern ORMs parameterize by default;
    //   recurs in raw SQL and AI-generated snippets) | legacy/10yr tech-debt HIGH.
    // EXAMPLE:
    //   GET /injection/sql/login?username=admin'--&password=whatever
    //   GET /injection/sql/login?username=x' OR '1'='1&password=x
    // FIX: parameterize (see /injection/sql/login-safe).
    // ========================================================================
    @GetMapping("/login")
    public Map<String, Object> login(@RequestParam(defaultValue = "") String username,
                                     @RequestParam(defaultValue = "") String password) {
        // VULNERABLE: user input concatenated straight into the SQL string.
        String sql = "SELECT id, username, role FROM users "
                + "WHERE username = '" + username + "' AND password = '" + password + "'";
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("query", sql);
        try {
            List<Map<String, Object>> rows = jdbc.queryForList(sql);
            out.put("authenticated", !rows.isEmpty());
            out.put("as", rows.isEmpty() ? null : rows.get(0));
        } catch (Exception e) {
            out.put("error", e.getMessage());
        }
        return out;
    }

    // ========================================================================
    // PERMUTATION 2 — SQL injection via String.format (UNION data theft)
    // OWASP 2021: A03 Injection      OWASP 2025: A05 Injection
    // CWE-89
    // EXPLOITATION LIKELIHOOD: HIGH
    //   A UNION SELECT reads arbitrary columns/tables (e.g. dump password
    //   hashes and SSNs from the users table) through a search box.
    // PREVALENCE TODAY: greenfield MEDIUM (raw SQL persists in search/reporting/
    //   analytics paths) | legacy/10yr tech-debt HIGH.
    // EXAMPLE:
    //   GET /injection/sql/search?q=zzz' UNION SELECT username, password, ssn FROM users--
    // FIX: parameterize the WHERE clause.
    // ========================================================================
    @GetMapping("/search")
    public Map<String, Object> search(@RequestParam(defaultValue = "") String q) {
        // VULNERABLE: String.format interpolation of untrusted input into SQL.
        String sql = String.format(
                "SELECT title, body, owner FROM notes WHERE body LIKE '%%%s%%'", q);
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("query", sql);
        try {
            out.put("results", jdbc.queryForList(sql));
        } catch (Exception e) {
            out.put("error", e.getMessage());
        }
        return out;
    }

    // ========================================================================
    // PERMUTATION 3 — SQL injection via a dynamic ORDER BY clause
    // OWASP 2021: A03 Injection      OWASP 2025: A05 Injection
    // CWE-89
    // EXPLOITATION LIKELIHOOD: MEDIUM
    //   Column/keyword positions cannot be bound as parameters, so developers
    //   often concatenate sort fields. Exploitable for blind boolean/error-based
    //   extraction; needs more skill than a UNION.
    // PREVALENCE TODAY: greenfield MED-HIGH (SQL identifiers/sort columns cannot be
    //   bound, so concatenated ORDER BY stays common in modern ORM apps) | legacy HIGH.
    // EXAMPLE:
    //   GET /injection/sql/notes?sort=body--
    //   GET /injection/sql/notes?sort=(SELECT CASE WHEN 1=1 THEN id ELSE owner END)
    // FIX: allow-list sortable column names.
    // ========================================================================
    @GetMapping("/notes")
    public Map<String, Object> notesSorted(@RequestParam(defaultValue = "id") String sort) {
        // VULNERABLE: sort field concatenated into ORDER BY.
        String sql = "SELECT id, owner, title FROM notes ORDER BY " + sort;
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("query", sql);
        try {
            out.put("results", jdbc.queryForList(sql));
        } catch (Exception e) {
            out.put("error", e.getMessage());
        }
        return out;
    }

    // ========================================================================
    // SAFE REFERENCE — parameterized query. Shown so the vulnerable/safe pair
    // can be diffed by SAST tooling and learners.
    // ========================================================================
    @GetMapping("/login-safe")
    public Map<String, Object> loginSafe(@RequestParam(defaultValue = "") String username,
                                         @RequestParam(defaultValue = "") String password) {
        List<Map<String, Object>> rows = jdbc.queryForList(
                "SELECT id, username, role FROM users WHERE username = ? AND password = ?",
                username, password);
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("authenticated", !rows.isEmpty());
        out.put("as", rows.isEmpty() ? null : rows.get(0));
        return out;
    }
}

File: csharp/Controllers/SqlInjectionController.cs
using Microsoft.AspNetCore.Mvc;

namespace VulnApp.Controllers;

/// <summary>
/// SQL Injection — OWASP 2021 A03 Injection | OWASP 2025 A05 Injection.
///
/// Reference controller for the C# app: mirror this annotation format and the
/// attribute-routed action style in every other C# vulnerability controller.
/// </summary>
[ApiController]
public class SqlInjectionController : ControllerBase
{
    // ========================================================================
    // PERMUTATION 1 — SQL injection via string interpolation (auth bypass)
    // OWASP 2021 A03 Injection      OWASP 2025 A05 Injection
    // CWE-89: Improper Neutralization of Special Elements used in an SQL Command
    // EXPLOITATION LIKELIHOOD: HIGH — pre-auth, deterministic, no tooling; the
    //   textbook  admin'--  /  ' OR '1'='1  payloads bypass the login.
    // PREVALENCE TODAY: greenfield LOW (EF Core parameterizes by default) |
    //   legacy/10yr tech-debt HIGH (hand-built ADO.NET/string SQL is common).
    // EXAMPLE: GET /injection/sql/login?username=admin'--&password=x
    // FIX: use parameters (see /injection/sql/login-safe).
    // ========================================================================
    [HttpGet("/injection/sql/login")]
    public IActionResult Login(string username = "", string password = "")
    {
        var sql = "SELECT id, username, role FROM users WHERE username=@u AND password=@p";
        try
        {
            var rows = Db.Query(sql, ("@u", username), ("@p", password));
            return new JsonResult(new { query = sql, authenticated = rows.Count > 0, @as = rows.Count > 0 ? rows[0] : null });
        }
        catch (Exception e)
        {
            return new JsonResult(new { query = sql, error = e.Message }) { StatusCode = 500 };
        }
    }

    // ========================================================================
    // PERMUTATION 2 — SQL injection via interpolation (UNION data theft)
    // OWASP 2021 A03 Injection      OWASP 2025 A05 Injection
    // CWE-89
    // EXPLOITATION LIKELIHOOD: HIGH — a UNION SELECT dumps password hashes and
    //   SSNs from the users table through a notes search box.
    // PREVALENCE TODAY: greenfield MEDIUM (raw SQL persists in reporting) |
    //   legacy/10yr tech-debt HIGH.
    // EXAMPLE: GET /injection/sql/search?q=zzz' UNION SELECT username, password, ssn FROM users--
    // FIX: parameterize the WHERE clause.
    // ========================================================================
    [HttpGet("/injection/sql/search")]
    public IActionResult Search(string q = "")
    {
        // VULNERABLE: interpolated LIKE clause.
        var sql = $"SELECT title, body, owner FROM notes WHERE body LIKE '%{q}%'";
        try
        {
            return new JsonResult(new { query = sql, results = Db.Query(sql) });
        }
        catch (Exception e)
        {
            return new JsonResult(new { query = sql, error = e.Message }) { StatusCode = 500 };
        }
    }

    // ========================================================================
    // PERMUTATION 3 — SQL injection via a dynamic ORDER BY clause
    // OWASP 2021 A03 Injection      OWASP 2025 A05 Injection
    // CWE-89
    // EXPLOITATION LIKELIHOOD: MEDIUM — identifiers can't be parameterized, so a
    //   concatenated sort field enables blind boolean/error-based extraction.
    // PREVALENCE TODAY: greenfield MED-HIGH (ORDER BY columns can't be bound) |
    //   legacy HIGH.
    // EXAMPLE: GET /injection/sql/notes?sort=body--
    // FIX: allow-list sortable column names.
    // ========================================================================
    [HttpGet("/injection/sql/notes")]
    public IActionResult Notes(string sort = "id")
    {
        // VULNERABLE: sort field concatenated into ORDER BY.
        var sql = $"SELECT id, owner, title FROM notes ORDER BY {sort}";
        try
        {
            return new JsonResult(new { query = sql, results = Db.Query(sql) });
        }
        catch (Exception e)
        {
            return new JsonResult(new { query = sql, error = e.Message }) { StatusCode = 500 };
        }
    }

    // ========================================================================
    // SAFE REFERENCE — parameterized query. Shown so the vulnerable/safe pair
    // can be diffed by SAST tooling and learners.
    // ========================================================================
    [HttpGet("/injection/sql/login-safe")]
    public IActionResult LoginSafe(string username = "", string password = "")
    {
        var rows = Db.Query("SELECT id, username, role FROM users WHERE username=@u AND password=@p",
            ("@u", username), ("@p", password));
        return new JsonResult(new { authenticated = rows.Count > 0, @as = rows.Count > 0 ? rows[0] : null });
    }
}
